File: PAT_Basic_Level_by_r/1071 小赌怡情/1071 小赌怡情.R
# 题目名称：小赌怡情
# 题目编号：PAT Basic 1071
# 实现原理：
#   模拟赌博游戏过程。初始给定筹码数，逐局处理游戏：
#   检查下注是否超过现有筹码，不足则提示；
#   否则根据赌大小结果增减筹码。
#   当筹码输光时游戏结束。
# 时间复杂度：O(K)，K为游戏局数
# 空间复杂度：O(1)，仅存储当前筹码数
x<-readLines("stdin",warn=FALSE);money<-as.integer(strsplit(x[1],"\\s+")[[1]][1]);for(s in x[2:length(x)]){v<-as.integer(strsplit(s,"\\s+")[[1]]);if(v[3]>money){cat("Not enough tokens.  Total = ",money,".\n",sep="");next};win<-(v[2]==0&&v[4]<v[1])||(v[2]==1&&v[4]>v[1]);money<-money+if(win)v[3]else-v[3];if(win)cat("Win ",v[3],"!  Total = ",money,".\n",sep="")else cat("Lose ",v[3],".  Total = ",money,".\n",sep="");if(money==0){cat("Game Over.\n");break}}

File: PAT_Basic_Level_by_perl/1071 小赌怡情/1071 小赌怡情.pl
use strict;
use warnings;

# =====================================================================
# 题目：1071 小赌怡情
# 难度：简单
# 知识点：模拟
# =====================================================================
#
# 【题目描述】
#   模拟下注游戏：每轮给出 n1 b t n2，玩家赌第二个数比第一个大还是小。
#   猜对赢得 t 筹码，猜错扣除 t 筹码；下注不能超过持有筹码；
#   筹码输光输出 Game Over 并结束。
#
# 【实现原理】
#   1. 若下注 t 大于持有筹码，输出 Not enough tokens 并跳过本轮。
#   2. 根据 b 与 n1、n2 的大小关系判断输赢并更新筹码。
#   3. 筹码归零时输出 Game Over 并结束。
#
# 【输入样例】
#   100 4
#   8 0 100 2
#   3 1 50 1
#   5 1 200 6
#   7 0 200 8
# 【输出样例】
#   Win 100!  Total = 200.
#   Lose 50.  Total = 150.
#   Not enough tokens.  Total = 150.
#   Not enough tokens.  Total = 150.
# =====================================================================

my $line = <STDIN>;
chomp $line;
my ($money, $k) = split /\s+/, $line;

for (1 .. $k) {
    my $l = <STDIN>;
    chomp $l;
    my ($n1, $b, $t, $n2) = split /\s+/, $l;

    if ($t > $money) {                       # 下注超过持有筹码
        print "Not enough tokens.  Total = $money.\n";
        next;
    }

    my $win = ($b == 1) ? ($n2 > $n1) : ($n2 < $n1);
    if ($win) {
        $money += $t;
        print "Win $t!  Total = $money.\n";
    } else {
        $money -= $t;
        print "Lose $t.  Total = $money.\n";
        if ($money == 0) {                   # 输光则游戏结束
            print "Game Over.\n";
            last;
        }
    }
}